frontend/src/data/saga.js

```diff
@@ -25,14 +26,40 @@ function* watchSingleRequest() {
 
 function* watchAddUser(action) {
   let { name } = action.payload;
+  yield performAsyncOperation(() => firebaseInteractor.addUser(name), "id");
+}
+
+function* watchDownloadImage(action) {
+  yield performAsyncOperation(
+    () => firebaseInteractor.downloadImage(action.payload.imageURL),
+    "imageURL"
+  );
+}
+
+/**
+ * Turns an async operation into a generator
+ * This at the end will put the `success` action into the value which will send the data through saga
+ *
+ * @param {Function<Void, Promise<T>>} operation the async operation to perform
+ * @param {String} payloadName the name of the object in the payload
+ * @template T
+ * Example:
+ * yield performAsyncOperation(() => new Promise((resolve, reject) => resolve(1)), "name")
+ * will generate a put(success({ name: 1 }))
+ */
+function* performAsyncOperation(operation, payloadName) {
+  let value;
+  const updateWithSuccess = async () => {
+    value = await operation();
+  };
   try {
-    let id;
-    const updateWithSuccess = async () => {
-      id = await firebaseInteractor.addUser(name);
-    };
     yield call(updateWithSuccess);
-    yield put(singleRequest.success({ id: id }));
+    // Create the object to return, using the payloadName and the returned value
+    let returnObject = {};
+    returnObject[payloadName] = value;
+    yield put(singleRequest.success(returnObject));
```

Comment:
I think this can just be singleRequest.success({ imageURL: value })

Reply:
This is in the `performAsyncOperation`, which is abstract over the payloadName. I wonder if the diff looks weird, I didn't actually check it.

Reply:
Oh interesting ill try to run it too

Reply:
I would say stay consistent? Cuz like in our success in action, it's doing object destructing. But ig if we have to do this then it's aight too

Reply:
I don't know of another way to have a string variable be the name of the argument. Is there a way to do it that I don't know?

Reply:
Oh wait I see what you are doing here cuz you have either id or imageURL. Nvm i think this is fine then lol. Potentially we can just create two success action since they are kinda doing different things

Reply:
So do you think this is a good abstraction, or should I go back to an unabstracted function? Which do you think is better?

Reply:
I would say to back to the unabstracted function :DD

Reply:
Back to unabstracted!. LMK what you think